File: leetcode/problem3170.cpp
//Lexicographically Minimum String After Removing Stars
#include <string>
#include <vector>
#include <iostream>
using namespace std;

class Solution {
public:
    string clearStars(string s) {
        int n = s.size();
        vector<bool> removed(n, false);
        vector<vector<int>> buckets(26);

        // 1) One pass: mark removals
        for (int i = 0; i < n; ++i) {
            if (s[i] == '*') {
                // remove the star itself
                removed[i] = true;
                // find & remove the smallest char to its left
                for (int c = 0; c < 26; ++c) {
                    if (!buckets[c].empty()) {
                        removed[buckets[c].back()] = true;
                        buckets[c].pop_back();
                        break;
                    }
                }
            } else {
                // record this letter's index
                buckets[s[i] - 'a'].push_back(i);
            }
        }

        // 2) Rebuild result from all indices not marked removed
        string result;
        result.reserve(n);  // at most n chars
        for (int i = 0; i < n; ++i) {
            if (!removed[i]) {
                result.push_back(s[i]);
            }
        }
        return result;
    }
};

// Add a main function for testing
int main() {
    Solution solution;
    string test = "a*b*c";
    cout << "Input: " << test << endl;
    cout << "Output: " << solution.clearStars(test) << endl;
    return 0;
}

File: leetcode/problem3170.py
#Lexicographically Minimum String After Removing Stars
class Solution:
    def clearStars(self, s: str) -> str:
        n=len(s)

        ans = list(s)

        buckets = [[]for _ in range(26)]
        for i,c in enumerate(s):
            if c == '*':
                #delete *
                ans[i]=''
                #find and delete the smallest non'*' character to its left
                for j in range(26):
                    #find if bucket[j] not a blank
                    if buckets[j]:
                        idx = buckets[j].pop()
                        ans[idx] = ''
                        break
            else:
                #append index of c
                buckets[ord(c)-ord('a')].append(i)
        return "".join(ans)
    
